From b09ab13fbf7458e9afde7dbfbc6f96a29691aa5b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Arkadiusz=20Mi=C5=9Bkiewicz?= <arekm@maven.pl>
Date: Thu, 22 Oct 2026 10:44:24 +0200
Subject: [PATCH] Fix #66049: Typemap can break parsing in parse_packet_soap
 leading to a segfault

A SOAP fault whose text decodes to a non-string (typemap callback, or
just xsi:type/xsi:nil from the server) is read via Z_STR(): segfault.

Origin: backport, https://github.com/php/php-src/commit/209f4c296ec6a08c721afdf17d787db4b5fd37d0
Bug: https://bugs.php.net/66049
PLD: 7.2: taken from packages/php branch PHP_7_1
---
 ext/soap/php_packet_soap.c                   |  3 ++
 ext/soap/tests/bugs/bug66049.phpt            | 48 ++++++++++++++++++++++++++
 ext/soap/tests/bugs/bug66049_no_typemap.phpt | 45 ++++++++++++++++++++++++
 3 files changed, 96 insertions(+)

diff --git a/ext/soap/php_packet_soap.c b/ext/soap/php_packet_soap.c
index 9da58ea1..cb133de5 100644
--- a/ext/soap/php_packet_soap.c
+++ b/ext/soap/php_packet_soap.c
@@ -195,6 +195,7 @@ int parse_packet_soap(zval *this_ptr, char *buffer, int buffer_size, sdlFunction
 			if (tmp != NULL && tmp->children != NULL) {
 				zval zv;
 				master_to_zval(&zv, get_conversion(IS_STRING), tmp);
+				convert_to_string(&zv);
 				faultstring = Z_STR(zv);
 			}
 
@@ -202,6 +203,7 @@ int parse_packet_soap(zval *this_ptr, char *buffer, int buffer_size, sdlFunction
 			if (tmp != NULL && tmp->children != NULL) {
 				zval zv;
 				master_to_zval(&zv, get_conversion(IS_STRING), tmp);
+				convert_to_string(&zv);
 				faultactor = Z_STR(zv);
 			}
 
@@ -225,6 +227,7 @@ int parse_packet_soap(zval *this_ptr, char *buffer, int buffer_size, sdlFunction
 				if (tmp != NULL && tmp->children != NULL) {
 					zval zv;
 					master_to_zval(&zv, get_conversion(IS_STRING), tmp);
+					convert_to_string(&zv);
 					faultstring = Z_STR(zv);
 				}
 			}
diff --git a/ext/soap/tests/bugs/bug66049.phpt b/ext/soap/tests/bugs/bug66049.phpt
new file mode 100644
index 00000000..0bbf449d
--- /dev/null
+++ b/ext/soap/tests/bugs/bug66049.phpt
@@ -0,0 +1,48 @@
+--TEST--
+Fix #66049 Typemap can break parsing in parse_packet_soap leading to a segfault
+--SKIPIF--
+<?php if (!extension_loaded('soap')) die('skip soap extension not available'); ?>
+--INI--
+soap.wsdl_cache_enabled=0
+--FILE--
+<?php
+function soap_string_from_xml($str)
+    { 
+    echo "soap_string_from_xml\n";
+
+	// Should return an string
+    return 2.3; 
+    }
+
+class TestSoapClient extends SoapClient {
+    function __doRequest($request, $location, $action, $version, $one_way = 0): ?string {
+        $res='<?xml version="1.0" encoding="UTF-8"?>
+            <SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/">
+            <SOAP-ENV:Body>
+            <SOAP-ENV:Fault><faultcode>SOAP-ENV:Server</faultcode><faultstring>not present</faultstring>
+            </SOAP-ENV:Fault></SOAP-ENV:Body></SOAP-ENV:Envelope>';
+        return $res;
+    }
+}
+
+try {
+    $client=new TestSoapClient(null, [
+        'uri' => 'test://',
+        'location' => 'test://',
+        'typemap' => [[
+            "type_ns" => "http://www.w3.org/2001/XMLSchema",
+            "type_name" => "string",
+            "from_xml" => "soap_string_from_xml"
+         ]]]);
+    $client->Mist("");
+} catch (SoapFault $e) {
+    var_dump($e->faultstring);
+    var_dump($e->faultcode);
+}
+?>
+Done
+--EXPECT--
+soap_string_from_xml
+string(3) "2.3"
+string(15) "SOAP-ENV:Server"
+Done
diff --git a/ext/soap/tests/bugs/bug66049_no_typemap.phpt b/ext/soap/tests/bugs/bug66049_no_typemap.phpt
new file mode 100644
index 00000000..1dfbfd2f
--- /dev/null
+++ b/ext/soap/tests/bugs/bug66049_no_typemap.phpt
@@ -0,0 +1,45 @@
+--TEST--
+Bug #66049 (non-string fault text from xsi:type or xsi:nil, no typemap needed)
+--SKIPIF--
+<?php if (!extension_loaded('soap')) die('skip soap extension not available'); ?>
+--INI--
+soap.wsdl_cache_enabled=0
+--FILE--
+<?php
+class TestSoapClient extends SoapClient {
+    public $res;
+    function __doRequest($request, $location, $action, $version, $one_way = 0) {
+        return $this->res;
+    }
+}
+
+$ns = 'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema"';
+$soap11 = '<?xml version="1.0"?><env:Envelope xmlns:env="http://schemas.xmlsoap.org/soap/envelope/" ' . $ns . '><env:Body><env:Fault>%s</env:Fault></env:Body></env:Envelope>';
+$soap12 = '<?xml version="1.0"?><env:Envelope xmlns:env="http://www.w3.org/2003/05/soap-envelope" ' . $ns . '><env:Body><env:Fault>%s</env:Fault></env:Body></env:Envelope>';
+
+$faults = array(
+    sprintf($soap11, '<faultcode>x</faultcode><faultstring xsi:type="xsd:int">5</faultstring><faultactor xsi:type="xsd:boolean">true</faultactor>'),
+    sprintf($soap11, '<faultcode>x</faultcode><faultstring xsi:nil="true">x</faultstring>'),
+    sprintf($soap12, '<env:Code><env:Value>env:Sender</env:Value></env:Code><env:Reason><env:Text xsi:type="xsd:int">7</env:Text></env:Reason>'),
+);
+
+$client = new TestSoapClient(null, array('location' => 'test://', 'uri' => 'test://'));
+foreach ($faults as $fault) {
+    $client->res = $fault;
+    try {
+        $client->test();
+    } catch (SoapFault $e) {
+        var_dump($e->faultstring);
+        if (isset($e->faultactor)) {
+            var_dump($e->faultactor);
+        }
+    }
+}
+?>
+Done
+--EXPECT--
+string(1) "5"
+string(1) "1"
+string(0) ""
+string(1) "7"
+Done
